# **Name: Subramanyam Raghuvamsidhar**
# **Mail ID: raghuvamsidhars@gmail.com**

In [ ]:
#Importing libraries
import pandas as pd
import numpy as np

In [ ]:
#Getting the datasets from sklearn library
from sklearn.datasets import load_iris
iris_datasets = load_iris()

In [ ]:
#printing the keys that are presemt in the dataset
print('keys of iris_datasets: \n{}'.format(iris_datasets.keys()))

keys of iris_datasets: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


In [ ]:
#This gives a basic description of the dataset that was provoded by sklearn library
print(iris_datasets['DESCR'][:193]+"\n...")

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, pre
...


In [ ]:
#This gives a basic idea of what are the types of flower are present
print('Target names:{}'.format(iris_datasets['target_names']))

Target names:['setosa' 'versicolor' 'virginica']


In [ ]:
#Printing the names of the features
print('Feature names: \n{}'.format(iris_datasets['feature_names']))


Feature names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [ ]:
#Now we are going to print the type of data
print('Type of Data:{}'.format(type(iris_datasets['data'])))

Type of Data:<class 'numpy.ndarray'>


As we came to a conclusion that now we are going to deal with ndarray 

In [ ]:
#Now we are going to know the shape of data or number of training examples
print("shape of data: {}".format(iris_datasets['data'].shape)) 

shape of data: (150, 4)


As we came to conclusion that we are going  to deal with 150 training examples

In [ ]:
#Printing the first 10 coloumns of the training data
print('First 10 coloumn of data is: {}'.format(iris_datasets['data'][:10]))

First 10 coloumn of data is: [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]


In [ ]:
#Printing the data type of target in the training data
print('Target type: {}'.format(type(iris_datasets['target'])))

Target type: <class 'numpy.ndarray'>


In [ ]:
#Now we are going to know the shape of target data
print("shape of target: {}".format(iris_datasets['target'].shape)) 

shape of target: (150,)


In [ ]:
#Printing the values of targets
print('Target: \n{}'.format(iris_datasets['target']))

Target: 
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [ ]:
#Splitting the data into training data and testing data by default it splits with 75:25 ratio
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(iris_datasets['data'],iris_datasets['target'],random_state = 0)

So finally we have splitted the iris data in train and test

In [ ]:
#Printing the shapes of splitted data
print('X_train shape: {}'.format(X_train.shape))
print('X_test shape: {}'.format(X_test.shape))
print('y_train shape: {}'.format(y_train.shape))
print('y_test shape: {}'.format(y_test.shape))

X_train shape: (112, 4)
X_test shape: (38, 4)
y_train shape: (112,)
y_test shape: (38,)


##Model Design and training


Here we are going to use K_Nearest_Neighbours algorithm and the reason for using that algo is by visualising the data that was plotted in the demo video. 

In [ ]:
#Using sklearn library we are going to use KNN algo
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 1) 

In [ ]:
knn.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

In [ ]:
#Lets make a random example to check wheather the model can predict or not
X_new = np.array([[4,1.9,2,0.4]])


#Lets make the prediction for X_new
Prediction = knn.predict(X_new)
print('Prediction:{}'.format(Prediction))
print("Prdicted type name is: {}".format(iris_datasets['target_names'][Prediction]))

Prediction:[0]
Prdicted type name is: ['setosa']


In [ ]:
#Now we are going to predict the values for test dataset
y_pred = knn.predict(X_test)
print('Test set predictions: \n{}'.format(y_pred))

Test set predictions: 
[2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [ ]:
#Printing the accuracy of our model on test dataset
print('Accuracy of prediction on test dataset:{}'.format(np.mean(y_pred==y_test)))

Accuracy of prediction on test dataset:0.9736842105263158


So, I got a accuracy of 97.3% on my test dataset which is very good.
